bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java

```diff
@@ -1281,68 +2122,259 @@ public static Iterable<HistoricItem> getAllStatesBetween(Item item, ZonedDateTim
+                first = new RetimedHistoricItem(first, beginTime);
+            }
+            if (first != null) {
+                betweenItemsList.add(0, first);
             }
         }
 
         // add HistoricItem at end
-        if (end != null && !end.isAfter(ZonedDateTime.now())) {
-            if (betweenItemsList.isEmpty()
-                    || !betweenItemsList.get(betweenItemsList.size() - 1).getTimestamp().equals(end)) {
-                HistoricItem last = historicState(item, end, serviceId);
+        if (betweenItemsList.isEmpty()
+                || !betweenItemsList.get(betweenItemsList.size() - 1).getTimestamp().equals(end)) {
+            HistoricItem last = endTime.equals(now) ? historicItemOrCurrentState(item, null)
+                    : internalPersistedState(item, endTime, serviceId);
+            if (last != null) {
+                last = new RetimedHistoricItem(last, endTime);
+            }
+            if (last != null) {
+                betweenItemsList.add(last);
+            }
+        }
+        return !betweenItemsList.isEmpty() ? betweenItemsList : null;
+    }
 
-                if (last != null) {
-                    betweenItemsList.add(new RetimedHistoricItem(last, end));
-                }
+    private static @Nullable PersistenceService getService(String serviceId) {
+        PersistenceServiceRegistry reg = registry;
+        return reg != null ? reg.get(serviceId) : null;
+    }
+
+    private static @Nullable String getDefaultServiceId() {
+        PersistenceServiceRegistry reg = registry;
+        if (reg != null) {
+            String id = reg.getDefaultId();
+            if (id != null) {
+                return id;
+            } else {
+                LoggerFactory.getLogger(PersistenceExtensions.class)
+                        .warn("There is no default persistence service configured!");
             }
+        } else {
+            LoggerFactory.getLogger(PersistenceExtensions.class)
+                    .warn("PersistenceServiceRegistryImpl is not available!");
         }
+        return null;
+    }
 
-        return betweenItemsList;
+    private static @Nullable DecimalType getItemValue(Item item) {
+        if (item instanceof NumberItem numberItem) {
+            Unit<?> unit = numberItem.getUnit();
+            if (unit != null) {
+                QuantityType<?> qt = item.getStateAs(QuantityType.class);
+                qt = (qt != null) ? qt.toUnit(unit) : qt;
+                if (qt != null) {
+                    return new DecimalType(qt.toBigDecimal());
+                }
+            }
+        }
+        return item.getStateAs(DecimalType.class);
     }
 
-    private static @Nullable HistoricItem historicItemOrCurrentState(Item item, HistoricItem historicItem,
-            DecimalType value) {
-        if (historicItem == null && value != null) {
+    private static @Nullable HistoricItem historicItemOrCurrentState(Item item, @Nullable HistoricItem historicItem) {
```

Comment:
This would be API breaking. HistoricItem a a State with a ZonedDateTime. It is returned when the ZonedDateTime makes sense. My original proposal was to use HistoricItem everywhere, but @J-N-K correctly reacted that it does not make sense for many of the methods. So I kept HistoricItem where it was already returned (and it makes sense) and use State where the ZonedDateTime does not make sense.

Reply:
No, I didn't mean to return HistoricItem everywhere. 

I mean that for methods that return a HistoricItem (such as maximumSince), or an Iterable<HistoricItem> (such as getAllStatesSince), make sure that the historicItem.getState() would return a QuantityType instead of DecimalType where applicable.

So from an API consumer's perspective:

```java
HistoricItem historyMax = PersistenceExtensions.maximumSince(myItem, ZonedDateTime.now().minusDays(1));
var max = historyMax.getState(); // <= this will be a QuantityType instead of DecimalType
```

Reply:
Unless I am missing something, I believe that is case already. There are also a series of tests testing these methods with QuantityTypes.

Reply:
My apologies, I didn't look at the tests, you're correct. Where did the QuantityType come from? Did it come from the persistence service?

Reply:
Yes, indeed.